File: Dart/lib/src/Classes/Result.dart
import 'ResultInfo.dart';

class Result<T> implements IResult<T> {
  Result(bool succeeded, T value, ResultInfo info) {
    Succeeded = succeeded;
    Value = value;
    Info = info;
  }

  Result.withoutValue(bool succeeded, ResultInfo info) {
    Succeeded = succeeded;
    Info = info;
  }

  Result.withoutResultInfo(bool succeeded, T value) {
    Succeeded = succeeded;
    Value = value;
  }

  @override
  ResultInfo Info;

  @override
  bool Succeeded;

  @override
  T Value;
}

abstract class IResult<T> {
  bool Succeeded;
  T Value;
  ResultInfo Info;
}

class ResultAPI {
  static IResult<T> Success<T>(T resValue) {
    return Result<T>(true, resValue, ResultInfo('No errors detected'));
  }

  static IResult<T> SuccessWithMessage<T>(String successMsg, T resValue) {
    return Result<T>(true, resValue, ResultInfo(successMsg));
  }

  static IResult<T> Fail<T>(String errMsg) {
    return Result<T>(false, null, ResultInfo(errMsg));
  }
}

File: Csharp/RadioJavan/Classes/Result.cs
using RadioJavan.Interfaces;
using System;
using System.Collections.Generic;
using System.Linq;
using System.Net.Http;
using System.Text;
using System.Threading.Tasks;

namespace RadioJavan.Classes
{
    public class Result<T> : IResult<T>
    {
        public Result(bool succeeded, T value, ResultInfo info)
        {
            Succeeded = succeeded;
            Value = value;
            Info = info;
        }

        public Result(bool succeeded, ResultInfo info)
        {
            Succeeded = succeeded;
            Info = info;
        }

        public Result(bool succeeded, T value)
        {
            Succeeded = succeeded;
            Value = value;
        }

        public bool Succeeded { get; }
        public T Value { get; }
        public ResultInfo Info { get; } = new ResultInfo("");
    }

    public static class Result
    {
        public static IResult<T> Success<T>(T resValue)
        {
            return new Result<T>(true, resValue, new ResultInfo("No errors detected"));
        }

        public static IResult<T> Success<T>(string successMsg, T resValue)
        {
            return new Result<T>(true, resValue, new ResultInfo(successMsg));
        }

        public static IResult<T> Fail<T>(Exception exception)
        {
            return new Result<T>(false, default(T), new ResultInfo(exception));
        }

        public static IResult<T> Fail<T>(string errMsg)
        {
            return new Result<T>(false, default(T), new ResultInfo(errMsg));
        }

        public static IResult<T> Fail<T>(string errMsg, T resValue)
        {
            return new Result<T>(false, resValue, new ResultInfo(errMsg));
        }

        public static IResult<T> Fail<T>(Exception exception, T resValue)
        {
            return new Result<T>(false, resValue, new ResultInfo(exception));
        }
        public static IResult<T> Fail<T>(ResultInfo info, T resValue)
        {
            return new Result<T>(false, resValue, info);
        }
    }
}
